pkg/controllers/provisioning/scheduling/scheduler.go

```diff
@@ -50,60 +50,70 @@ func init() {
 }
 
 type Scheduler struct {
-	kubeClient client.Client
-	topology   *Topology
+	kubeClient  client.Client
+	preferences *Preferences
 }
 
 func NewScheduler(kubeClient client.Client) *Scheduler {
 	return &Scheduler{
-		kubeClient: kubeClient,
-		topology:   &Topology{kubeClient: kubeClient},
+		kubeClient:  kubeClient,
+		preferences: NewPreferences(),
 	}
 }
 
-func (s *Scheduler) Solve(ctx context.Context, provisioner *v1alpha5.Provisioner, instanceTypes []cloudprovider.InstanceType, pods []*v1.Pod) ([]*Node, error) {
+func (s *Scheduler) Solve(ctx context.Context, provisioner *v1alpha5.Provisioner, instanceTypes []cloudprovider.InstanceType, pods []*v1.Pod) (nodes []*Node, err error) {
 	defer metrics.Measure(schedulingDuration.WithLabelValues(injection.GetNamespacedName(ctx).Name))()
 	constraints := provisioner.Spec.Constraints.DeepCopy()
 
 	sort.Slice(pods, byCPUAndMemoryDescending(pods))
 	sort.Slice(instanceTypes, byPrice(instanceTypes))
 
-	// Inject temporarily adds specific NodeSelectors to pods, which are then
-	// used by scheduling logic. This isn't strictly necessary, but is a useful
-	// trick to avoid passing topology decisions through the scheduling code. It
-	// lets us treat TopologySpreadConstraints as just-in-time NodeSelectors.
-	if err := s.topology.Inject(ctx, constraints, pods); err != nil {
-		return nil, fmt.Errorf("injecting topology, %w", err)
-	}
+	topology := NewTopology(s.kubeClient, constraints)
```

Comment:
I think eventually we're gonna want to inject this into the scheduler and make it longlived.

Reply:
It will make inverse anti-affinities less expensive, but definitely for another PR due to the complexity of ensuring we track correctly. It may get complicated as well as we have to undo those topology changes if nodes fail to launch 😆

Reply:
We will definitely need to be able to update the guessed state of the world with the actual. I'm imagining this looks exactly the same as updating w/ actual and just overwrite keys.